### DISCUSSION TITLE
Parallel access and synching of devices

### DISCUSSION THREAD

**Discussion by pyZerrenner** [pymeasure/pymeasure#1064] [Q&A] [answered]
The question #916 already touched on this, but the main topic was not really answered: Is there a intended way to access multiple devices in parallel in a procedure (or multiple procedures) and potentially synch their sampling times? Or is that entirely on me, in how I write my procedure (e.g. using `multiprocessing`)? 

As an example: in my setup, I have a RS485 port with two measurement cards (those I have to access in sequence), an USB measurement card and potentially a camera. It would be good to get the data from these devices (RS485, USB card, camera) in parallel to speed up the acquisition. I specifically say "device" and not "measurement", since (as discussed in #916) accessing the same device in multiple threads is generally complicated.

I assume that these issues will be solved by [LECO](https://github.com/pymeasure/leco-protocol) in the future. Is [PyLECO](https://github.com/pymeasure/pyleco) already stable enough to tested for this?

**Comment by BenediktBurger** [comment, marked as answer]
> Is there a intended way to access multiple devices in parallel in a procedure (or multiple procedures) and potentially synch their sampling times?

It depends on your requirements for "parallel sampling":
- The intended way is to read one instrument after another.
- If the device communication introduces to much lag, you could use multi-threading (no need to use multi-processing, as the bottle neck is the device) to start the devices more or less at the same time.
- As an alternative, you can use pyleco as a form or multi-threading/processing
- If your requirements are quite strict (a few milliseconds), you enter the _real-time-computing_ area, which probably requires dedicated hardware, as the different message buffers and operating system process handling introduce not calculable delays.

> Is [PyLECO](https://github.com/pymeasure/pyleco) already stable enough to tested for this?

My whole lab is based on pyleco and its running smoothly quite some time, at least the last six months.

**Reply by pyZerrenner** [reply]
Thanks for the answer. My target sampling rates are currently slower than 10 Hz, so the timing does not have to be extremely precise. 

I'll have a look at pyleco. But before that, would it make sense to use the Procedure->Results->Worker structure of PyMeasure within my main Procedure to handle the multithreading? Something along the lines of
```python
class myProcedure(Procedure):
    # treat this as a naive "pseudocode"
    def execute(self):
        device1_procedure = Procedure()  # write a separate procedure for each device
        device1_results = Results(device1_procedure , data1_filename)
        device1_worker = Worker(device1_results )

        # do the same setup for device2 and device3

        device1_worker.start()
        device2_worker.start()
        device3_worker.start()
```
Or is nesting of Procedures/Workers not a good idea and this should rather be done by using the `threading` module directly?

**Reply by BenediktBurger** [reply]
For each data point, you have to do a "result.emit" with the data of all three devices.
Therefore, you have to do a single Procedure, as far as I know.
Sou you would have to use threading.

I'm not an expert in the pymeasure data acquisition, as I have my own data acquisition scheme (which evolved into pyleco). Maybe @msmttchr  or @CasperSchippers  do know more.

**Reply by CasperSchippers** [reply]
At a certain point I did something that sounds somewhat similar to what you want. In that case the issue was that there were sevaral instruments that required communicating at different moments in time (and at very different rates), so slightly different but can maybe serve as inspiration still. I'll try to explain how I set it up (and maybe there are other/better approaches, but this is how I thought to fix the issue).

Basically I had a [procedure](https://github.com/CasperSchippers/spynwave/blob/master/src/spynwave/procedures/procedure_field_sweep.py), which started a [thread for every instrument](https://github.com/CasperSchippers/spynwave/blob/master/src/spynwave/procedures/threads.py). These threads would sent some commands (in my case this meant setting magnetic field values or sweeping a frequency) and also acquire the data. Whenever data was acquired, it was put on a Queue to a [different, data-collecting thread](https://github.com/CasperSchippers/spynwave/blob/master/src/spynwave/drivers/data_thread.py) that would collect the data from all the instrument-threads (i.e., emptying all the Queues of the different threads) and match the datapoints (based on time-stamp). It also needed to average a bunch of datapoints since one of the instruments measuring at a very different frequency that the other instruments such that there would be single value per datapoint. This thread would then emit the data (for all instruments) to get it written to file & plotted.

For your case, if you want to poll all instruments at the same time/rate, I think you could drop the 'data-collecting' thread and just launch a bunch of instrument-threads from your procedure. Then in a for-loop (or something to control how often your measurement is performed and when it is stopped) you could ask the threads to perform a measurement (via an event or a queue of something), collect the data by emptying all the queues, and emit the data.

**Reply by pyZerrenner** [reply]
Thanks for your replies. So I take from it, that parallel processing within a procedure has to be done by hand. I mainly wanted to avoid missing some build-in way of pymeasure to handle this case.

**Reply by CasperSchippers** [reply]
That is indeed correct, there is no ready solution for this. I do find the  `StoppableThread` (or the gui version `StoppableQThread`), `InterruptableEvent` provided in pymeasure very handy, since that does take away a bit of the boilerplate and fixes some issues (like not being able to be stopped in a manner that is convenient for instruments and the like) with the  'normal' versions.
